File: ros2_project/src/2-1TopicCommunication/one-way_topic_communication_cpp/src/GetHello.cpp
#include <std_msgs/msg/string.hpp>
#include <functional>
#include <rclcpp/logging.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/subscription.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>

class GetHello : public rclcpp::Node{
public:
    GetHello(std::string name) : rclcpp::Node(name){
        // 创建一个订阅者
        get_hello = this->create_subscription<std_msgs::msg::String>("/HelloWorld",10,std::bind(&GetHello::SubHello,this,std::placeholders::_1));
    }

private:
    // 订阅者所需的回调函数
    void SubHello(std_msgs::msg::String msg){
        RCLCPP_INFO(this->get_logger(),"订阅者接收到消息：%s",msg.data.c_str());
    }

    // 声明一个订阅者指针
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr get_hello;
};

int main(int argc, char ** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GetHello>("SubHello");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_project/src/2-1TopicCommunication/one_way_topic_communication_py/one_way_topic_communication_py/GetHello.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class GetHello(Node):
    # 构造函数，在每个类对象初始化时都会执行
    def __init__(self,name):
        super().__init__(name)

        # 创建一个订阅者（Python中不需要声明）
        self.get_hello = self.create_subscription(String,"/HelloWorld",self.get_hello_callback,10)

    # 订阅者回调函数（这里msg:String其实是规定变量msg必须是std_msgs.msg.String类型）
    def get_hello_callback(self,msg:String):
        # 注意：因为std_msgs.msg.String里面封装了一个str类型的data属性，所以我们实际上是操纵这个data属性
        self.get_logger().info(f"订阅者接收到消息: {msg.data}")

def main(args=None):
    rclpy.init(args=args)       # 初始化rclpy
    node = GetHello("SubHello") # 新建一个节点
    rclpy.spin(node)            # 保持节点运行，检测是否收到退出指令(Ctrl+C)
    rclpy.shutdown()            # 当接收到退出指令时，退出spin函数，执行shutdown函数来关闭rclpy

